src/main/java/org/apache/brooklyn/cloudfoundry/entity/VanillaPaasApplicationCloudFoundryDriver.java

```diff
@@ -0,0 +1,207 @@
+                                                    CloudFoundryPaasLocation location) {
+        this.entity = checkNotNull(entity, "entity");
+        this.location = checkNotNull(location, "location");
+        applicationName = entity.getApplicationName();
+    }
+
+    @Override
+    public VanillaCloudFoundryApplicationImpl getEntity() {
+        return entity;
+    }
+
+    @Override
+    public CloudFoundryPaasLocation getLocation() {
+        return location;
+    }
+
+    @Override
+    public void start() {
+        deploy();
+        preLaunch();
+        launch();
+        postLaunch();
+    }
+
+    private String deploy() {
+        Map<String, Object> params = MutableMap.copyOf(entity.config().getBag().getAllConfig());
+        params.put(VanillaCloudFoundryApplication.APPLICATION_NAME.getName(), applicationName);
+
+
+        if (!Strings.isBlank(entity.getConfig(VanillaCloudFoundryApplication.ARTIFACT_PATH))) {
+            params.put(VanillaCloudFoundryApplication.ARTIFACT_PATH.getName(), getLocalPath());
+        }
+        applicationUrl = location.deploy(params);
+        return applicationUrl;
+    }
+
+    private String getLocalPath() {
+        DownloadResolver downlodResolver = getDownloadResolver();
+        try {
+            File war;
+            war = LocalResourcesDownloader
+                    .downloadResourceInLocalDir(downlodResolver.getFilename(),
+                            downlodResolver.getTargets());
+            return war.getCanonicalPath();
+        } catch (IOException e) {
+            log.error("Error obtaining local path in {} for artifact {}",
+                    this, downlodResolver.getTargets());
+            throw Exceptions.propagate(e);
+        }
+    }
+
+    private DownloadResolver getDownloadResolver() {
+        String artifactUrl = entity.getConfig(VanillaCloudFoundryApplication.ARTIFACT_PATH);
+        return new BasicDownloadResolver(ImmutableList.of(artifactUrl),
+                FileNameResolver.findArchiveNameFromUrl(artifactUrl));
+    }
+
+    protected void preLaunch() {
+        configureEnv();
+    }
+
+    protected void configureEnv() {
+        setEnv(entity.getConfig(VanillaCloudFoundryApplication.ENV));
+    }
+
+    @Override
+    public void setEnv(Map<String, String> env) {
+        if ((env != null) && (!env.isEmpty())) {
+            location.setEnv(applicationName, env);
+        }
+        entity.sensors().set(VanillaCloudFoundryApplication.APPLICATION_ENV,
+                location.getEnv(applicationName));
+    }
+
+    private void launch() {
+        location.startApplication(applicationName);
+    }
+
+    private void postLaunch() {
+        entity.sensors().set(Attributes.MAIN_URI, URI.create(applicationUrl));
+        entity.sensors().set(VanillaCloudFoundryApplication.ROOT_URL, applicationUrl);
+        updateMemorySensor(entity.getConfig(VanillaCloudFoundryApplication.REQUIRED_MEMORY));
```

Comment:
In fact, I was thinking about that. First, I used the API to set the value to the sensor. But after I decided to use the config instead, because the resources (memory, disk, and instances) value were just configured in the previous step (application creation). So, I thought the application configuration was very near and in `postlaunch` method we could assume that the application resource values in the cloud were the same that the config keys, so I did not need a API called.
But, I should not assume anything, so I agree with you, the sensor should be feed by the API.